plugins/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/DatabaseReaderLazyLoader.java

```diff
@@ -27,30 +28,116 @@
     }
 
-    synchronized DatabaseReader get() throws IOException {
+    /**
+     * Read the database type from the database. We do this manually instead of relying on the built-in mechanism to avoid reading the
+     * entire database into memory merely to read the type. This is especially important to maintain on master nodes where pipelines are
+     * validated. If we read the entire database into memory, we could potentially run into low-memory constraints on such nodes where
+     * loading this data would otherwise be wasteful if they are not also ingest nodes.
+     *
+     * @return the database type
+     * @throws IOException if an I/O exception occurs reading the database type
+     */
+    final String getDatabaseType() throws IOException {
+        if (databaseType.get() == null) {
+            synchronized (databaseType) {
+                if (databaseType.get() == null) {
+                    final long fileSize = databaseFileSize();
+                    if (fileSize <= 512) {
+                        throw new IOException("unexpected file length [" + fileSize + "] for [" + databasePath + "]");
+                    }
+                    final int[] databaseTypeMarker = {'d', 'a', 't', 'a', 'b', 'a', 's', 'e', '_', 't', 'y', 'p', 'e'};
+                    try (InputStream in = databaseInputStream()) {
+                        // read the last 512 bytes
+                        final long skipped = in.skip(fileSize - 512);
+                        if (skipped != fileSize - 512) {
+                            throw new IOException("failed to skip [" + (fileSize - 512) + "] bytes while reading [" + databasePath + "]");
+                        }
+                        final byte[] tail = new byte[512];
+                        int read = 0;
+                        do {
+                            final int actualBytesRead = in.read(tail, read, 512 - read);
+                            if (actualBytesRead == -1) {
+                                throw new IOException("unexpected end of stream [" + databasePath + "] after reading [" + read + "] bytes");
+                            }
+                            read += actualBytesRead;
+                        } while (read != 512);
+
+                        // find the database_type header
+                        int metadataOffset = -1;
+                        int markerOffset = 0;
+                        for (int i = 0; i < tail.length; i++) {
+                            byte b = tail[i];
+
+                            if (b == databaseTypeMarker[markerOffset]) {
+                                markerOffset++;
+                            } else {
+                                markerOffset = 0;
+                            }
+                            if (markerOffset == databaseTypeMarker.length) {
+                                metadataOffset = i + 1;
+                                break;
+                            }
+                        }
+
```

Comment:
maybe add:

```java
if (metadataOffset == -1) {
   throw new IOException("database type marker not found")
}
```

Otherwise it would fail with an ArrayOutOfBoundException.

Reply:
I pushed 5a0ee8fbf54549f6bc75e1903326eb5c4ea8f032.